rust/src/lib/ovn.rs

```diff
@@ -0,0 +1,268 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{Deserialize, Deserializer, Serialize};
+use std::collections::HashMap;
+use std::fmt;
+use std::str::FromStr;
+
+pub const OVN_BRIDGE_MAPPINGS: &str = "ovn-bridge-mappings";
+
+#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize)]
+#[non_exhaustive]
+pub struct OvnConfiguration {
+    #[serde(
+        rename = "bridge-mappings",
+        skip_serializing_if = "Option::is_none"
+    )]
+    pub bridge_mappings: Option<Vec<OvnBridgeMapping>>,
+    #[serde(skip)]
+    pub(crate) prop_list: Vec<&'static str>,
+}
+
+impl OvnConfiguration {
+    pub fn is_none(&self) -> bool {
+        self.bridge_mappings.is_none()
+    }
+}
+
+impl<'de> Deserialize<'de> for OvnConfiguration {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let mut ovn_conf = Self::default();
+        let v = serde_json::Value::deserialize(deserializer)?;
+
+        if let Some(v) = v.as_object() {
+            if let Some(v) = v.get("bridge-mappings") {
```

Comment:
If you are doing `Deserialize` by yourself, you need to handle invalid key like:

```
ovn:
  invalid-set:
    abc: 1
  bridge-mappings:
    - localnet: yellow
      abc: ab
      bridge: ovsbr3
```

The `#[serde(deny_unknown_fields)]` can do the magic if using auto deserializer `#[derive(Deserialize)]`. From my view, you do not need special deserializer here.

Reply:
OK, I'll throw away the custom deserializer and run with derived.

Reply:
Still trying to do this ... Facing some issues.

Reply:
I'm thinking about this some more and there may be a reason to use a custom deserializer - although not a very good one.

If I use a custom deserializer I can provide a better error msg to the user if they try to provision the following bridge mapping:
```yaml
ovn:
  bridge-mappings:
  - localnet: net1
```

The above example is wrong because the user **must** also provide the bridge.

Right now, all I get is:
```
[2023-07-07T07:37:51Z INFO  nmstate::query_apply::net_state] Retrying on: VerificationError: Verification failure: ovn.bridge-mappings desire '[{"localnet":"net1"}]', current '[]'
```

Thoughts @cathay4t ?

Reply:
The `sanitize()` function is designed post-deserialize check and clean up.

Reply:
Added a `sanitize` func to ensure a mappings being added / modified has the required info.

Also moved asserting there aren't duplicate `localnet` keys in the desired state to a separate sanitize func.

Finally, removed the custom deserializer.